### PR TITLE
Add some tests for compact serialization coverage API-1790

### PR DESCRIPTION
**Pull request by srknzl**
Add tests for schema different than CompactWriter, field does not exist or different type. See https://hazelcast.atlassian.net/browse/API-1790 for more.

There are three points in the jira issue: 

1. I added test for this one.
2. I added test for this one.
3. It was already covered. 

Breaking changes (list specific methods/types/messages):
- none

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible
- [x] Send backports/forwardports if fix needs to be applied to past/future releases
- [x] New public APIs have `@Nonnull/@Nullable` annotations
- [x] New public APIs have `@since` tags in Javadoc

### CODE REVIEW

File: hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactSerializationTest.java [hunk 1]
```diff
@@ -430,6 +439,125 @@ public void testReflectiveSerializer_withFieldsWithOtherSerializationMechanisms_
         assertTrue(data.isCompact());
     }
 
+    @Test
+    public void testCompactWriterThrowsExceptionWhileWritingWhenSchemaDoesNotHaveThatField() {
+        thrown.expect(HazelcastSerializationException.class);
+        thrown.expectMessage("Failed to serialize");
+        thrown.expectCause(new RootCauseMatcher(HazelcastSerializationException.class, "Invalid field name"));
```
Comment (mdumandag) [inline hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactSerializationTest.java:446]: Personally, I don't like this style of asserting exceptions are thrown. What if any line other than the last one throws this? I believe the test will pass again, despite us expecting it to throw in the last line. I would suggest using assertThatThrownBy for the last line. The same applies to other test as well

File: hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactSerializationTest.java [hunk 2]
```diff
@@ -430,6 +439,125 @@ public void testReflectiveSerializer_withFieldsWithOtherSerializationMechanisms_
...
+        schemaService.put(schema);
+        CompactSerializationConfig compactSerializationConfig = new CompactSerializationConfig();
+        compactSerializationConfig.addSerializer(serializer);
+        SerializationService serializationService = new DefaultSerializationServiceBuilder()
+                .setSchemaService(schemaService)
+                .setConfig(new SerializationConfig().setCompactSerializationConfig(compactSerializationConfig))
+                .build();
+
+        // Add correct schema to classToSchemaMap
+        serializationService.toData(fooBar);
```
Comment (mdumandag) [inline hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactSerializationTest.java:465]: Do we need to add schema to the schema service above? This line would do it as well, right?

**Comment by srknzl** [comment]
Coverage before (current master): 

image


Coverage after: 

image



Reports:
 
Master:
image

PR:

image





**TLDR:** Coverage increased for compact package from 83.7 to 83.9 percent

**Comment by srknzl** [comment]
For those who want the most detail, reports are attached.

[htmlReportMaster.zip](https://github.com/hazelcast/hazelcast/files/10568611/htmlReportMaster.zip)
[htmlReportPR.zip](https://github.com/hazelcast/hazelcast/files/10568612/htmlReportPR.zip)

File: hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactSerializationTest.java (hunk 1 above)
Comment (srknzl) [reply hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactSerializationTest.java:446]: Ok makes sense, I was not used to it. Let me try to switch to it.

File: hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactSerializationTest.java (hunk 2 above)
Comment (srknzl) [reply hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactSerializationTest.java:465]: Correct, removing

File: hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactSerializationTest.java [hunk 3]
```diff
@@ -430,6 +433,120 @@ public void testReflectiveSerializer_withFieldsWithOtherSerializationMechanisms_
...
+        FooBarSerializer serializer = spy(new FooBarSerializer());
+        FooBar fooBar = new FooBar("foo", 42L);
+
+        SchemaService schemaService = CompactTestUtil.createInMemorySchemaService();
+        CompactSerializationConfig compactSerializationConfig = new CompactSerializationConfig();
+        compactSerializationConfig.addSerializer(serializer);
+        SerializationService serializationService = new DefaultSerializationServiceBuilder()
+                .setSchemaService(schemaService)
+                .setConfig(new SerializationConfig().setCompactSerializationConfig(compactSerializationConfig))
+                .build();
```
Comment (mdumandag) [inline hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactSerializationTest.java:447]:
```suggestion
        SerializationService serializationService = createSerializationService(() -> serializer);
```
can be simplified to this

File: hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactSerializationTest.java [hunk 4]
```diff
@@ -430,6 +433,120 @@ public void testReflectiveSerializer_withFieldsWithOtherSerializationMechanisms_
...
+        Schema schema = schemaWriter.build();
+
+        SchemaService schemaService = CompactTestUtil.createInMemorySchemaService();
+        schemaService.put(schema);
+        CompactSerializationConfig compactSerializationConfig = new CompactSerializationConfig();
+        compactSerializationConfig.addSerializer(serializer);
+        SerializationService serializationService = new DefaultSerializationServiceBuilder()
+                .setSchemaService(schemaService)
+                .setConfig(new SerializationConfig().setCompactSerializationConfig(compactSerializationConfig))
+                .build();
```
Comment (mdumandag) [inline hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactSerializationTest.java:484]:
```suggestion
        SerializationService serializationService = createSerializationService(() -> serializer);
```

this one can also be simplified

File: hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactSerializationTest.java (hunk 3 above)
Comment (srknzl) [reply hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactSerializationTest.java:447]: 👍🏻

**Review by mdumandag — APPROVED** [review]

**Review by emreyigit — APPROVED** [review]
